core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1035,11 +1035,13 @@ class ReplicaManager(val config: KafkaConfig,
       val partitionFetchSize = fetchInfo.maxBytes
       val followerLogStartOffset = fetchInfo.logStartOffset
 
-      brokerTopicStats.topicStats(tp.topic).totalFetchRequestRate.mark()
-      brokerTopicStats.allTopicsStats.totalFetchRequestRate.mark()
-
       val adjustedMaxBytes = math.min(fetchInfo.maxBytes, limitBytes)
       try {
+        brokerTopicStats.allTopicsStats.totalFetchRequestRate.mark()
+        if (allPartitions.contains(tp)) {
```

Comment:
Hmm.. Does this solve the issue or just make it less likely? Does anything protect a call to `stopReplica` between this check and the metric update below?

It does seem this is related to #8586. Since we are updating the metric after completion of the `DelayedFetch` currently, then this case actually seems likely to be hit in practice. We would just need to have a fetch in purgatory when the call to stop replica is received. However, after #8586, I guess it becomes less likely? Still it would be nice to think through a complete fix.

Reply:
Yeah, this is the reason this PR got stuck a bit. @rajinisivaram had said:

>I had a look through the code and I think it should be possible to close the timing window altogether by creating metrics within one of the existing locks (locking should be required only when metrics for the topic is not available). I am not sure if it is worth the additional complexity.

Reply:
I think the problem here is that the metric is created on demand. I think we need to tie it to partition lifecycles a little more closely. The thought I had is basically just to create the topic metric whenever we receive a LeaderAndIsr request so that creation/deletion are protected by `replicaStateChangeLock`. We can then just ignore updates to the metric if it doesn't exist rather than letting it be recreated.

Reply:
This PR makes the issue less likely to happen. It's not bulletproof but we've been using this patch in production for years and has worked great for us. Heuristic and pragmatic...